### PR TITLE
Override toxworkdir with --workdir.

### LINKED ISSUE #2654: --workdir override not working in tox >=4
## Issue

In tox <4 using --workdir would override whatever was in tox.ini, this no longer works in >= 4 which makes some CI/CD use cases difficult 

## Environment

Provide at least:

- OS: linux

## Minimal example

If possible, provide a minimal reproducer for the issue:

```
tox.ini:
[tox]
toxworkdir=/tmp/blah

[testenv]
allowlist_externals=echo
commands=echo {toxworkdir}

setup.py
from distutils.core import setup

setup(name='Distutilsaaa',
      version='1.0',
      description='Python Distribution Utilities',
      author='Greg Ward',
      author_email='gward@python.net',
      url='https://www.python.org/sigs/distutils-sig/',
      packages=['distutilsa'],
     )

distutilsa/__init.py

```

Tox 4:
```
 jabbera    toxbug   3.8.10  ﮫ20.419s⠀   tox -e py --workdir /tmp/blah1  70  19:30:47 
.pkg: install_requires> python -I -m pip install 'setuptools>=40.8.0' wheel
.pkg: _optional_hooks> python /home/jabbera/.local/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
.pkg: get_requires_for_build_sdist> python /home/jabbera/.local/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
.pkg: prepare_metadata_for_build_wheel> python /home/jabbera/.local/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
.pkg: build_sdist> python /home/jabbera/.local/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
py: install_package> python -I -m pip install --force-reinstall --no-deps /tmp/blah/.tmp/package/1/Distutilsaaa-1.0.tar.gz
py: commands[0]> echo /tmp/blah
/tmp/blah
.pkg: _exit> python /home/jabbera/.local/lib/python3.10/site-packages/pyproject_api/_backend.py True setuptools.build_meta __legacy__
  py: OK (6.34=setup[6.34]+cmd[0.00] seconds)
  congratulations :) (6.43 seconds)
```

Tox 3:
```
 jabbera    toxbug   3.8.10  ﮫ4.404s⠀   tox -e py --workdir /tmp/blah1         bash   70  19:29:36 
GLOB sdist-make: /home/jabbera/toxbug/setup.py
py inst-nodeps: /tmp/blah1/.tmp/package/1/Distutilsaaa-1.0.zip
py installed: Distutilsaaa @ file:///tmp/blah1/.tmp/package/1/Distutilsaaa-1.0.zip
py run-test-pre: PYTHONHASHSEED='1612454640'
py run-test: commands[0] | echo /tmp/blah1
/tmp/blah1
_______________________________________________________ summary ________________________________________________________
  py: commands succeeded
  congratulations :)
```

### PR DESCRIPTION
**Pull request by q0w**
# Thanks for contribution
Closes #2654 
Please, make sure you address all the checklists (for details on how see
[development documentation](http://tox.readthedocs.org/en/latest/development.html#development))!

- [x] ran the linter to address style issues (`tox -e fix`)
- [x] wrote descriptive pull request text
- [x] ensured there are test(s) validating the fix
- [x] added news fragment in `docs/changelog` folder
- [ ] updated/extended the documentation

### CODE REVIEW

File: tests/config/test_sets.py [hunk 1]
```diff
@@ -171,3 +172,10 @@ def test_set_env_raises_on_non_str(mocker: MockerFixture) -> None:
     env_set.loaders.insert(0, MemoryLoader(set_env=1))
     with pytest.raises(TypeError, match="1"):
         assert env_set["set_env"]
+
+
+def test_config_work_dir(tox_project: ToxProjectCreator) -> None:
+    project = tox_project({"tox.ini": "[tox]\ntoxworkdir=/tmp/foo"})
```
Comment (gaborbernat) [inline tests/config/test_sets.py:178]: This path seems invalid on Windows can we just do foo instead?

File: tests/config/test_sets.py [hunk 2]
```diff
@@ -171,3 +172,10 @@ def test_set_env_raises_on_non_str(mocker: MockerFixture) -> None:
...
+    work_dir_flag = "a"
+    result = project.run("c", "-k", "toxworkdir", "--core", "--workdir", work_dir_flag)
```
Comment (gaborbernat) [inline tests/config/test_sets.py:180]: Can we use the primary key for this instead?

**Comment by q0w** [comment]
> Can we use the primary key for this instead?
how?

**Comment by q0w** [comment]
@gaborbernat should work_dir end with .tox?

**Comment by gaborbernat** [comment]
Only by default, if the user sets a dir we should use that.

File: src/tox/config/sets.py [hunk 3]
```diff
@@ -191,10 +191,14 @@ def register_config(self) -> None:
         def work_dir_builder(conf: Config, env_name: str | None) -> Path:  # noqa: U100
             return (conf.work_dir if conf.work_dir is not None else cast(Path, self["tox_root"])) / ".tox"
 
+        def work_dir_post_process(value: Path) -> Path:
+            return self._conf.work_dir if self._conf.work_dir is not None and self._conf.options.work_dir else value
```
Comment (q0w) [inline src/tox/config/sets.py:195]: so here i dont need to add .tox to work_dir?

File: tests/config/test_sets.py [hunk 4]
```diff
@@ -171,3 +171,11 @@ def test_set_env_raises_on_non_str(mocker: MockerFixture) -> None:
...
     with pytest.raises(TypeError, match="1"):
         assert env_set["set_env"]
+
+
+@pytest.mark.parametrize("work_dir", ["a", ""])
+def test_config_work_dir(tox_project: ToxProjectCreator, work_dir: str) -> None:
+    project = tox_project({"tox.ini": "[tox]\ntoxworkdir=b"})
+    result = project.run("c", *(["--workdir", work_dir] if work_dir else []))
+    expected = Path(project.path, work_dir) if work_dir else Path("b")
+    assert expected == result.state.conf.core["toxworkdir"]
```
Comment (gaborbernat) [inline tests/config/test_sets.py:181]: Use work_dir instead of toxworkdir

**Review by gaborbernat — CHANGES_REQUESTED** [review]

File: tests/config/test_sets.py (hunk 4 above)
Comment (q0w) [reply tests/config/test_sets.py]: tests are still failing on windows

**Review by gaborbernat — APPROVED** [review]

**Pull request merged**